controllers/etcd_controller.go

```diff
@@ -159,30 +179,159 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	logger.Infof("Succesfully reconciled etcd: %s", etcd.GetName())
+
+	return ctrl.Result{
+		Requeue:      true,
+		RequeueAfter: time.Second * 5,
+	}, nil
+}
+
+func (r *EtcdReconciler) getEtcdService(etcd *druidv1alpha1.Etcd) (*corev1.ServiceList, error) {
+	etcdServiceList := &corev1.ServiceList{}
+	err := r.List(context.TODO(), etcdServiceList, client.InNamespace(etcd.Namespace), client.MatchingLabels(etcd.Spec.Labels))
+	return etcdServiceList, err
+}
+
+func (r *EtcdReconciler) getEtcdStatefulSet(etcd *druidv1alpha1.Etcd) (*appsv1.StatefulSetList, error) {
+	etcdStatefulSetList := &appsv1.StatefulSetList{}
+	err := r.List(context.TODO(), etcdStatefulSetList, client.InNamespace(etcd.Namespace), client.MatchingLabels(etcd.Spec.Labels))
+	return etcdStatefulSetList, err
+}
+
+func (r *EtcdReconciler) getEtcdConfigMap(etcd *druidv1alpha1.Etcd) (*corev1.ConfigMapList, error) {
+	etcdConfigMapList := &corev1.ConfigMapList{}
+	err := r.List(context.TODO(), etcdConfigMapList, client.InNamespace(etcd.Namespace), client.MatchingLabels(etcd.Spec.Labels))
+	return etcdConfigMapList, err
+}
+
+func (r *EtcdReconciler) getExistingEtcdResourcesAndSetOwnerRef(etcd *druidv1alpha1.Etcd) (bool, error) {
+	etcdServiceList, err := r.getEtcdService(etcd)
+	if err != nil {
+		return false, err
+	}
+
+	etcdStatefulsetList, err := r.getEtcdStatefulSet(etcd)
+	if err != nil {
+		return false, err
+	}
+
+	etcdConfigMapList, err := r.getEtcdConfigMap(etcd)
+	if err != nil {
+		return false, err
+	}
+
+	if (len(etcdServiceList.Items) + len(etcdStatefulsetList.Items) + len(etcdConfigMapList.Items)) == 0 {
+		return false, nil
+	}
+
+	for _, serviceToPatch := range etcdServiceList.Items {
+		service := serviceToPatch.DeepCopy()
+		if !checkForEtcdOwnerReference(serviceToPatch.OwnerReferences, etcd) {
+			if err := controllerutil.SetControllerReference(etcd, &serviceToPatch, r.Scheme); err != nil {
+				return true, err
+			}
+			err := r.Patch(context.TODO(), &serviceToPatch, client.MergeFrom(service))
+			if err != nil {
+				return true, err
+			}
+		}
+	}
+
+	for _, ssToPatch := range etcdStatefulsetList.Items {
+		ss := ssToPatch.DeepCopy()
+		if !checkForEtcdOwnerReference(ssToPatch.OwnerReferences, etcd) {
+			if err := controllerutil.SetControllerReference(etcd, &ssToPatch, r.Scheme); err != nil {
+				return true, err
+			}
+			err := r.Patch(context.TODO(), &ssToPatch, client.MergeFrom(ss))
+			if err != nil {
+				return true, err
+			}
+		}
+	}
+
+	for _, cmToPatch := range etcdConfigMapList.Items {
+		cm := cmToPatch.DeepCopy()
+		if !checkForEtcdOwnerReference(cmToPatch.OwnerReferences, etcd) {
+			if err := controllerutil.SetControllerReference(etcd, &cmToPatch, r.Scheme); err != nil {
+				return true, err
+			}
+			err := r.Patch(context.TODO(), &cmToPatch, client.MergeFrom(cm))
+			if err != nil {
+				return true, err
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func checkForEtcdOwnerReference(refs []metav1.OwnerReference, etcd *druidv1alpha1.Etcd) bool {
```

Comment:
This checks only if any of the refs have the UID of etcd. This is necessary but [not sufficient](https://stupefied-goodall-e282f7.netlify.com/contributors/design-proposals/api-machinery/controller-ref/#adoption). We should not adopt objects with refs to some other controller.

It would be better to rename this func as `getAdoptionState(refs []metav1.OwnerReference, etcd *druidv1alpha1.Etcd) adoptionState` with values `ADOPTED`, `CAN_ADOPT`, `CANNOT_ADOPT`.

Reply:
I have changed the implementation logic of adoption. PTAL.